### PR TITLE
Operator overloading implemented

### PR DESCRIPTION
**Pull request by hgupta-tavisca**
Please review the code
@ytokas

### CODE REVIEW

File: Training/01 Operator Overloading/Operator Overloading.Host/Program.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Operator_Overloading.Model;
+
+namespace Operator_Overloading.Host
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Host/Program.cs]: no special chars in namespace

File: Training/01 Operator Overloading/Operator Overloading.Host/Program.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+           double  amountFirst,amountSecond;
+            string currencyFirst,currencySecond;
+            Console.WriteLine("Enter the first amount : ");
+            amountFirst = Convert.ToDouble(Console.ReadLine());
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Host/Program.cs]: look at double.TryParse

File: Training/01 Operator Overloading/Operator Overloading.Model/MismatchException.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿
+
+using System;
+namespace Operator_Overloading.Model
+{
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/MismatchException.cs]: Make it serializable

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+using System.Threading.Tasks;
+
+namespace Operator_Overloading.Model
+{
+    public class Money
+    {
+        private double _amount;
+        private string _currency;
+
+        public double Amount
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: public double Amount {get; set;}

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                return this._currency;
+            }
+            set
+            {
+                this._currency = value;
+            }
+        }
+
+        public Money(double Amount1, string Currency)  // constructor
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: why Amount1?

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            if (Amount1 < 0.0)
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: Amount can be negative

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                throw new ArgumentException();  // throw exception if amount is negative
+            }
+            else
+            {
+                _amount = Amount;
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: this.Amount = Amount

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        public static Money operator +(Money amountOne, Money amountTwo)
+        {
+            Money money3 = new Money();
+           
+           
+            
+            
+                if (amountOne.Currency == amountTwo.Currency)
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: use string.Equals and check for case sensitivity

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                {
+                    money3.Amount = amountOne.Amount + amountTwo.Amount;
+                    if (money3.Amount == 0.0 && amountOne.Amount != (-amountTwo.Amount))
```
Comment (hgupta-tavisca) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: ??????????????????????????????????????

File: Training/01 Operator Overloading/Operator Overloading.Host/Program.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        static void Main(string[] args)
+        {
+            double amountFirst,amountSecond,value;
+            string currencyFirst,currencySecond;
+            string value1, value2;
+            Console.WriteLine("Enter the first amount : ");
+            value1 = Console.ReadLine();
+            if (Double.TryParse(value1, out value))
+                amountFirst = value;
+            else
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Host/Program.cs]: Why is amount 0 if parsing fails? If input is incorrect ask user to enter again.

File: Training/01 Operator Overloading/Operator Overloading.Host/Program.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                amountFirst = 0;
+          
+            Console.WriteLine("Enter Currency : ");
+            currencyFirst = Console.ReadLine();
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Host/Program.cs]: Add null checks for currency. Also check string length should be 3 char

File: Training/01 Operator Overloading/Operator Overloading.Host/Program.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            currencySecond = Console.ReadLine();
+            Money money1 = new Money(amountFirst, currencyFirst);
+            Money money2 = new Money(amountSecond, currencySecond);
+            Money money3 = new Money();
+            try
+            {
+                 money3 = money1 + money2;
+            }
+            catch (MismatchException misMatchException)
+            {
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Host/Program.cs]: Why create new property in custom exception? Use message property of the Exception class.

File: Training/01 Operator Overloading/Operator Overloading.Model/MismatchException.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.IO;
+using System.Xml;
+using System.Xml.Serialization;
+namespace OperatorOverloading.Model
+{
+     [Serializable()]
+    public class MismatchException : Exception
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/MismatchException.cs]: make it serializable, adding serializable attribute doesn't make it serializable, you need to implement proper constructor.

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        public string Currency
+        {
+            get;
+
+            set;
+
+        }
+
+        public Money(double Amount, string Currency)  // constructor
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: variables should be camel casing

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public static Money operator +(Money amountOne, Money amountTwo)
+        {
+            Money money3 = new Money();
+
+            
+            
+            
+                if (amountOne.Currency.Equals( amountTwo.Currency))
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: Add null checks for money objects and their currencies.
Check for case sensitivity

File: Training/01 Operator Overloading/Operator Overloading.Model/Money.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+                {
+                    money3.Amount = amountOne.Amount + amountTwo.Amount;
+                    if (money3.Amount == 0.0 && amountOne.Amount != (-amountTwo.Amount))
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/Money.cs]: did you test this???
try using double.isinfinity()

File: Training/01 Operator Overloading/Operator Overloading.Model/NotIntException.cs [hunk 17]
```diff
@@ -0,0 +1,27 @@
+﻿
+
+using System;
+namespace OperatorOverloading.Model
+{
+
+
+    public class  NotIntException : Exception
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/NotIntException.cs]: make it serializable

File: Training/01 Operator Overloading/Operator Overloading.Model/NotIntException.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        private string _NotIntExceptionInfo;
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/NotIntException.cs]: private variables should be in camelcasing preceded with '_'

File: Training/01 Operator Overloading/Operator Overloading.Model/OverflowException.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
+﻿
+
+using System;
+namespace OperatorOverloading.Model
+{
+
+
+    public class OverflowExceptionOverloaded : Exception
```
Comment (ytokas) [inline Training/01 Operator Overloading/Operator Overloading.Model/OverflowException.cs]: make it serializable, change the name of the Exception as told to you in the last review.

**Comment by ytokas** [comment]
Remove bin and obj folders and commit

**Pull request merged**